### PR TITLE
Components: Add a WAI-ARIA compliant custom select.

### LINKED ISSUE #16473: Investigate creating an accessible custom select/dropdown component
**Is your feature request related to a problem? Please describe.**

HTML `select` elements are not visually customisable, and in some circumstances we need to customise the options in a dropdown. When showing available text styles, for example:

Screen Shot 2019-07-09 at 2 09 38 pm

We also need these dropdowns to be fully accessible when used with keyboard, screen readers, voice commands, and any assistive technology. 

**Describe the solution you'd like**

Investigate how to best achieve full accessibility while using non-semantic markup, with ARIA and custom keyboard interaction. 
Previous discussion for reference: 
https://github.com/WordPress/gutenberg/pull/16148

Deque's custom select might be a good solution: https://pattern-library.dequelabs.com/components/selects

React-select is another possibility; there's ongoing work/discussion on accessibility improvements over there: https://github.com/JedWatson/react-select/issues/2456

### PR DESCRIPTION
**Pull request by epiqueras**
Closes #16473

## Description

This PR implements a fully accessible and customizable select component with individually style-able items, in the `@wordpress/components` package.

It uses [Downshift](https://github.com/downshift-js/downshift), a library that will be very useful for any of our other complex a11y input needs.

It has some modifications hooked on top of it to make it fully compliant with [the spec](https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html). These modifications might soon make it into a future [Downshift](https://github.com/downshift-js/downshift) release and at that point we should remove them from our implementations.

**The story for the component is:**

https://deploy-preview-17926--gutenberg-playground.netlify.com/design-system/components/?path=/story/customselect--default

## How has this been tested?

It was verified that the component works as expected.

Although the libraries used are heavily tested, unit testing this implementation once the approach is approved wouldn't be a bad idea to make sure we never regress to configuring something incorrectly.

## Screenshots

Screen Shot 2019-10-13 at 12 46 18 PM

## Types of Changes

*New Feature:* `@wordpress/components` now has a fully WAI-ARIA compliant custom select.

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/components/src/custom-select/index.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+// This is needed so that in Windows, where
+// the menu does not necessarily open on
+// key up/down, you can still switch between
+// options with the menu closed.
+const stateReducer = (
+	{ selectedItem },
+	{ type, changes, props: { items } }
+) => {
+	switch ( type ) {
+		case Downshift.stateChangeTypes.ToggleButtonKeyDownArrowDown:
```
Comment (silviuavram) [inline packages/components/src/custom-select/index.js]: it's actually `useSelect.stateChangeTypes` here and below.

File: packages/components/src/custom-select/index.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+ * Internal dependencies
+ */
+import { Button, Dashicon } from '../';
+
+const itemToString = ( item ) => item.name;
+// This is needed so that in Windows, where
+// the menu does not necessarily open on
+// key up/down, you can still switch between
+// options with the menu closed.
+const stateReducer = (
```
Comment (silviuavram) [inline packages/components/src/custom-select/index.js]: related to the Dragon Speech issue, I can look into it. if you are faster at it, I would suggest adding a console log here in the stateReducer and check the `type`, `changes` and `state`. If there is a state change in the select, then it should pass here.

**Comment by youknowriad** [comment]
Design-wise the options should be shown in a Popover I think.

File: packages/components/src/custom-select/index.js (hunk 1 above)
Comment (epiqueras) [reply packages/components/src/custom-select/index.js]: Thank you!

File: packages/components/src/custom-select/index.js (hunk 2 above)
Comment (epiqueras) [reply packages/components/src/custom-select/index.js]: Great idea!

**Comment by epiqueras** [comment]
>Design-wise the options should be shown in a Popover I think.

I want to tick all the `a11y` boxes first, and then see what we can do with styling without regressing.

File: packages/components/src/custom-select/index.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+import { useSelect } from 'downshift';
+import classnames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import { Button, Dashicon } from '../';
+
+const itemToString = ( item ) => item.name;
+// This is needed so that in Windows, where
```
Comment (tellthemachines) [inline packages/components/src/custom-select/index.js]: Just checked this on Windows and the behaviour is identical to Mac: on arrow down/up the menu opens. I don't see that as a major issue though. If we did want the arrow keys to work with the menu closed we'd probably have to make sure the options always render, and hide them with CSS, but then it's not clear that we could limit that behaviour to Windows 🤷‍♀

File: packages/components/src/custom-select/index.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+		delete menuProps[ 'aria-activedescendant' ];
+	}
+	return (
+		<div className="components-custom-select">
+			{ /* eslint-disable-next-line jsx-a11y/label-has-associated-control, jsx-a11y/label-has-for */ }
+			<span
+				{ ...getLabelProps( { className: 'components-custom-select__label' } ) }
+			>
+				{ label }
+			</span>
```
Comment (tellthemachines) [inline packages/components/src/custom-select/index.js]: Any reason not to make this an actual `<label>` instead of using `aria-labelledby`? Then we might not need the `aria-label`. Having both, VoiceOver and NVDA are announcing the button as "Font size font size". Would be good to avoid the repetition.

**Review by tellthemachines — COMMENTED** [review]
Good work on this! Checked with VO/Safari and NVDA/Firefox and it's behaving pretty well on both, apart from the repetition of "Font size" label I commented on below.

One thing I noticed on NVDA but was unable to reproduce consistently is that sometimes the font size button is announced as "expanded" when focused, and with the dropdown closed. I realise this may not be very helpful, will try and debug further 😄 

Another potential issue is that focus is not being limited to the component when it is open, so if we open the dropdown and then press tab we move out of it but the dropdown stays open. Shift-tabbing until the font size button is focused again allows us to continue interacting with the dropdown, but after that pressing `esc` will not close it. 

Also, minor styling issue on Firefox and IE:

Screen Shot 2019-10-18 at 11 36 46 am

Removing `display: inline-flex` and adding `text-align: left` to `components-button` will fix it for both browsers 🙂

File: packages/components/src/custom-select/index.js (hunk 3 above)
Comment (silviuavram) [reply packages/components/src/custom-select/index.js]: the issue here is the returned `changes`. @epiqueras just return the `selectedItem` because if you also return the rest of the `changes` then one of them is `isOpen` that comes as true :D

File: packages/components/src/custom-select/index.js (hunk 4 above)
Comment (silviuavram) [reply packages/components/src/custom-select/index.js]: The way I wrote `useSelect` was with having a physical label that also labels the button and the menu. As I understood there was an issue with Dragon Speech because of this.

Having said that, anyone tried Dragonn Speech to select an item then check what the console.log prints?

File: packages/components/src/custom-select/index.js (hunk 3 above)
Comment (epiqueras) [reply packages/components/src/custom-select/index.js]: Fixed :smile:

File: packages/components/src/custom-select/index.js (hunk 4 above)
Comment (epiqueras) [reply packages/components/src/custom-select/index.js]: >Any reason not to make this an actual <label> instead of using aria-labelledby? Then we might not need the aria-label. Having both, VoiceOver and NVDA are announcing the button as "Font size font size". Would be good to avoid the repetition.

Dragon Speech still needs the explicit aria label on the toggle to be able to click on it.

>Having said that, anyone tried Dragonn Speech to select an item then check what the console.log prints?

It registers a click on the toggle button I think:

Screen Shot 2019-10-18 at 1 58 17 PM

In that test^^ "Small" was already the selected item.

**Comment by epiqueras** [comment]
@tellthemachines Thanks for the help!

>Removing display: inline-flex and adding text-align: left to components-button will fix it for both browsers 🙂

Done :smile:

**Review by tellthemachines — CHANGES_REQUESTED** [review]
We're getting there! 

One remaining issue, apart from the comments I left: pressing Tab when the dropdown is open still causes focus to be moved to the button control. Subsequently pressing Esc will _not_ close the dropdown, but pressing any arrow key _will_ close it.

Also, would be good to test replacing the current font size picker with this component, to check focus management when entering/exiting it.

**Comment by enriquesanchez** [comment]
I ran some tests of the component in https://deploy-preview-17926--gutenberg-playground.netlify.com/design-system/components/?path=/story/customselect--default with Dragon, NVDA and Voice Over.

Here are the videos of the test:

**With Dragon: https://cldup.com/cihauaUTN4.m4v**
So far with Dragon I haven't able to select an option from the list by saying "Click Large", but I'm able to do so if I use the "move up/down" commands to switch focus to an option and then say "press enter".  I'm a fairly new user of speech recognition software, so I'm not sure how common or uncommon this flow is. That being said, I was able to interact with the component and select an option.

**With NVDA: https://cldup.com/7n79VbZMo3.m4v** (sorry for the shaky camera)
Besides "Font Size" being announced repetitively, the only thing that I felt was missing was that on Windows I can't change the selection by using the up/down keys, like on Mac.

**With Voice Over: https://cldup.com/A12FmcNrSn.m4v**
Same as with NVDA, "Font Size" is announced repetitively. Othe than that, interaction was as expected.

**Comment by tellthemachines** [comment]
> Besides "Font Size" being announced repetitively, the only thing that I felt was missing was that on Windows I can't change the selection by using the up/down keys, like on Mac.

@enriquesanchez this happens because NVDA is not switching into forms mode when entering this component. If you manually switch by pressing Insert+Space then you'll be able to use up/down keys. It would be worth investigating if it's possible to make NVDA switch automatically as that would be closer in behaviour to a native control.

**Comment by adamsilverstein** [comment]
Wanted to mention previous work to use Downshift for the post author selector in https://github.com/WordPress/gutenberg/pull/7478. It might be worth reviewing the comments on that ticket - ultimately, the accessibility team rejected this approach and we settled instead on using `accessible-autocomplete` - see https://github.com/WordPress/gutenberg/pull/7385.

@enriquesanchez - had you reviewed this previous work? What do you think of Downshift vs. accessible-autocomplete?

**Comment by silviuavram** [comment]
Adam can you mention the reasons for making that decision? I am interested
for improvement reasons at least. Thanks!

File: packages/components/src/custom-select/index.js (hunk 4 above)
Comment (epiqueras) [reply packages/components/src/custom-select/index.js]: Done!

File: packages/components/src/custom-select/style.scss [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+.components-custom-select__label {
+	display: block;
+	margin-bottom: 5px;
+}
+
+.components-custom-select__button {
+	border: 1px solid $dark-gray-200;
+	border-radius: 4px;
+	color: $dark-gray-500;
+	display: initial;
```
[the remark this answers is not in the document]
Comment (epiqueras) [reply packages/components/src/custom-select/style.scss]: Done!

**Comment by epiqueras** [comment]
I've made some optimizations and replaced the font size picker's select with this one.

Can we move forward with this PR and iterate in following ones now?

@tellthemachines @enriquesanchez @youknowriad

**Comment by tellthemachines** [comment]
> Adam can you mention the reasons for making that decision? I am interested for improvement reasons at least. Thanks!

@silviuavram [this comment](https://github.com/WordPress/gutenberg/pull/7478#issuecomment-400599022) explains the reasons, though that was over a year ago so perhaps some of those issues have been fixed by now.

File: packages/components/src/custom-select/index.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+						'screen-reader-text': hideLabelFromVision,
+					} ),
+				} ) }
+			>
+				{ label }
+			</label>
+			<Button
+				{ ...getToggleButtonProps( {
+					// This is needed because some speech recognition software don't support `aria-labelledby`.
+					'aria-label': label,
```
Comment (tellthemachines) [inline packages/components/src/custom-select/index.js]: The aria-label on the button will need to be the same as the visible label, because that's what speech to text users will try to activate the button with. In the previous iteration @enriquesanchez and I settled on making the labels for the font-size controls visible and instead hiding the "font size" fieldset legend:

Screen Shot 2019-09-25 at 6 33 31 PM

**Review by tellthemachines — CHANGES_REQUESTED** [review]
This is getting really close! 

I left some comments below, and besides that, there's a couple issues still needing attention. The dropdown is currently pushing the whole sidebar content up on desktop:
font-picker-jump
This can be fixed by setting the overflow on `.components-panel` to `visible`, which doesn't seem to break anything else, but may need a little extra testing. It looks _really_ broken now though, so worth giving it a go 😅 

The other issue is with the `aria-labelledby` on the button. I'd mentioned previously that both VoiceOver and NVDA were reading out the button label repeatedly, and have now figured out that that is happening because `aria-labelledby` is pointing to the button's own id, and because the button text is overridden by its `aria-label` this means that essentially `aria-label` is read out twice.
Screen Shot 2019-11-06 at 4 44 15 pm
If we remove the reference to the button's id in that `aria-labelledby` the problem will be solved. This seems to be set in Downshift, is there any way to override it? Cc @silviuavram would be good to know why it was originally implemented in this way 🙂

**Comment by silviuavram** [comment]
@tellthemachines yes, it comes from Downshift. I followed the pattern from [the ARIA website](https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html).

It doesn't make sense to me to have a labelledby if the screen reader already knows how to read the text from it. It does not really help if the button has no text either.

I can add a PR in Downshift for this tomorrow, since I am at WebConference right now. Or you can do that and I will approve tomorrow. The labelledby should probably point only to the label. What do you think?

**Comment by epiqueras** [comment]
@tellthemachines Thanks for the thorough reviews and fixes :smile:! It should be good to go now.

>The labelledby should probably point only to the label. What do you think?

We don't need it in this case, because we explicitly set an `aria-label` which has better support from voice control software. It was double labelled, once by the `aria-label` and once by the `aria-labelledby`. I cleared the `aria-labelledby`.

**Comment by silviuavram** [comment]
Oh ok, I understand. Interesting why a native button with text would ever need an aria-label. :(

**Comment by epiqueras** [comment]
Voice control software doesn't follow the spec :cry:

**Comment by enriquesanchez** [comment]
I did a quick test using Voice Control on the Mac:

voiceControl 2019-11-15 17_53_55

[Here's the video recording](https://cldup.com/f0JbC7gh0m.m4v) 

Based on the little experience I've had with it, Voice Control is a bit different that Dragon. Manipulation is done via a grid and numbers, so labels have no effect here.

That being said, I was able to interact with the component and select different font sizes.

**Comment by epiqueras** [comment]
Can we get another review please :smile:

cc @youknowriad @tellthemachines @gziolo

**Comment by tellthemachines** [comment]
The labelling is behaving much better now! 
For clarity though, because "font size" is the fieldset legend that applies to all 3 fields inside it, each field should have its own label. That will help for Dragon users to be able to interact with them too. @enriquesanchez and I discussed this previously and came up with this design:

Screen Shot 2019-09-25 at 6 33 31 PM

Where 
- the "font size" fieldset legend is visually hidden, because sighted users should be able to grasp what those fields are about from the surrounding context;
- "Size preset" (or "Preset size", which sounds better?) is the visible label for the picker, and
- "Custom" is the visible label for the range control.

There's also a little weirdness with the alignment of the options; the checked one is bumped slightly to the left due to the negative margin on the SVG. But that could be left for a follow-up PR 😅

**Comment by epiqueras** [comment]
@tellthemachines Thanks for taking another look! :smiley:

I think "Preset Size" sounds better. It makes more sense for "Size" to be the noun since its the more important word in this context.

I implemented that design and fixed the weirdness in the alignment with the check mark.

**Review by tellthemachines — APPROVED** [review]

**Comment by epiqueras** [comment]
Just in time for the :turkey: :tada: !

**Comment by youknowriad** [comment]
Nice work here :) A goodd follow-up would be to check where we're using `per_page: -1` in REST API calls and replace them with a paginated Select.

**Comment by epiqueras** [comment]
>Nice work here :) A good follow-up would be to check where we're using per_page: -1 in REST API calls and replace them with a paginated Select.

It's on my list :smile:

File: packages/edit-post/src/components/sidebar/style.scss [hunk 7]
```diff
@@ -25,7 +26,7 @@
 		position: relative;
 
 		@include break-small() {
-			overflow: hidden;
+			overflow: visible;
 			height: auto;
 			max-height: none;
 		}
```
Comment (youknowriad) [inline packages/edit-post/src/components/sidebar/style.scss]: Not sure I understand why the changes in this file are needed for this particular PR?

File: packages/components/src/index.js [hunk 8]
```diff
@@ -16,6 +16,7 @@ export { default as ClipboardButton } from './clipboard-button';
 export { default as ColorIndicator } from './color-indicator';
 export { default as ColorPalette } from './color-palette';
 export { default as ColorPicker } from './color-picker';
+export { default as CustomSelect } from './custom-select';
```
Comment (youknowriad) [inline packages/components/src/index.js]: Why it's named CustomSelect and not just Select

File: packages/edit-post/src/components/sidebar/style.scss (hunk 7 above)
Comment (epiqueras) [reply packages/edit-post/src/components/sidebar/style.scss]: https://github.com/WordPress/gutenberg/pull/17926#pullrequestreview-312188131

Comment (youknowriad) [reply packages/edit-post/src/components/sidebar/style.scss]: Why it doesn't just use Popover, this issue is already solved there, and I think the changes here might cause more unexpected bugs (like the scrolling issue we noticed)

File: packages/components/src/index.js (hunk 8 above)
Comment (epiqueras) [reply packages/components/src/index.js]: We already have `SelectControl` so that felt confusing, but I don't have a strong opinion on it.

File: storybook/test/__snapshots__/index.js.snap [hunk 9]
```diff
@@ -1539,61 +1539,78 @@ exports[`Storyshots Components|FontSizePicker Default 1`] = `
 <fieldset
   className="components-font-size-picker"
 >
-  <legend>
+  <legend
```
Comment (youknowriad) [inline storybook/test/__snapshots__/index.js.snap]: useless snapshots that we change as the implementation is updated?

File: packages/edit-post/src/components/sidebar/style.scss (hunk 7 above)
Comment (epiqueras) [reply packages/edit-post/src/components/sidebar/style.scss]: Is the `Popover` fully accessible?

cc @tellthemachines @enriquesanchez

File: packages/components/src/font-size-picker/index.js [hunk 10]
```diff
@@ -51,38 +57,40 @@ function FontSizePicker( {
...
 						className={ 'components-font-size-picker__select' }
-						label={ 'Choose preset' }
-						hideLabelFromVision={ true }
-						value={ currentSelectValue }
-						onChange={ onSelectChangeValue }
-						options={ getSelectOptions( fontSizes ) }
+						label={ __( 'Preset Size' ) }
+						items={ items }
+						selectedItem={ items.find( ( item ) => item.key === currentSelectValue ) || items[ 0 ] }
+						onSelectedItemChange={ onSelectChangeValue }
```
Comment (youknowriad) [inline packages/components/src/font-size-picker/index.js]: Why not just onChange (I'm trying to see if we can mirror the SelectControl API props, so it's consistent and switching between the two becomes easy.

File: storybook/test/__snapshots__/index.js.snap (hunk 9 above)
Comment (epiqueras) [reply storybook/test/__snapshots__/index.js.snap]: Yes, from the "storyshots".

File: packages/components/src/index.js (hunk 8 above)
Comment (youknowriad) [reply packages/components/src/index.js]: Good remark, I do think in that case that it should be SomethingControl. `CustomSelectControl` maybe. I don't like Custom that much though but I also don't have a lot of ideas.

File: packages/edit-post/src/components/sidebar/style.scss (hunk 7 above)
Comment (youknowriad) [reply packages/edit-post/src/components/sidebar/style.scss]: If it's not it should be :) but we did a lot of a11y work on Popover back then.

File: packages/components/src/index.js (hunk 8 above)
Comment (epiqueras) [reply packages/components/src/index.js]: I prefer `Select` to be honest. The goal is for it to be THE React version of `select` so that makes sense.

`Control` makes it seem like it uses `BaseControl` which it doesn't.

File: packages/components/src/font-size-picker/index.js (hunk 10 above)
Comment (epiqueras) [reply packages/components/src/font-size-picker/index.js]: I tried to mirror the `Downshift` props.

Their reasoning is probably that you also have `highlightedItem`, `onHighlightedItemChange`, etc. They prefer to be explicit about the two instead of having `selectedItem == value`.

Comment (youknowriad) [reply packages/components/src/font-size-picker/index.js]: I have a preference for staying consistent across our components, Downshift is just an implementation detail that can be changed at any time.

Comment (epiqueras) [reply packages/components/src/font-size-picker/index.js]: I think that having `onChange` and `onHighlightedValueChange` is a bit confusing. It's not totally clear that `onChange` does not include highlighted value changes.

Also, `selectedItem` takes the whole object, not just the key.

Similarly, `onSelectedItemChange` gives you the whole object, not just the key.

This already makes them fundamentally different to `SelectControl`'s `value` and `onChange`, but they are a much simpler to use API as you don't have to handle all of the key-to-object normalization/searching on the consumer side.

File: packages/edit-post/src/components/sidebar/style.scss (hunk 7 above)
Comment (enriquesanchez) [reply packages/edit-post/src/components/sidebar/style.scss]: I'm not aware of any accessibility issues with `Popover`.

File: packages/components/src/font-size-picker/index.js (hunk 10 above)
Comment (youknowriad) [reply packages/components/src/font-size-picker/index.js]: I feel like highlighting is not something you'd need to care about often as a consumer, I'm not sure we even need it as an API in our component to be honest.

Also, I just noticed that the component is lacking a README.

File: storybook/test/__snapshots__/index.js.snap (hunk 9 above)
Comment (tellthemachines) [reply storybook/test/__snapshots__/index.js.snap]: Why useless? I'd expect snapshots to change as the markup changes. I'd also expect to review those changes and make sure they are consistent with what we were trying to do, and nothing changed that shouldn't. That's the whole purpose of snapshots 🙂

File: packages/components/src/index.js (hunk 8 above)
Comment (tellthemachines) [reply packages/components/src/index.js]: I think it would be good to keep the "Custom", or some similar indication that this is not a standard HTML select (which we should also continue to provide, for cases where the options don't need to be visually distinguishable from one another)

File: packages/edit-post/src/components/sidebar/style.scss (hunk 7 above)
Comment (tellthemachines) [reply packages/edit-post/src/components/sidebar/style.scss]: There are no issues that I know of with `Popover` itself, but we had some trouble with focus management when using it in our previous attempt to implement this component: https://github.com/WordPress/gutenberg/pull/17418#pullrequestreview-294158647

Comment (epiqueras) [reply packages/edit-post/src/components/sidebar/style.scss]: Yes, that's what I was thinking of.

Part of the reason I didn't use `Popover` here from the start is that previous attempts had failed with it.

File: packages/components/src/font-size-picker/index.js (hunk 10 above)
Comment (epiqueras) [reply packages/components/src/font-size-picker/index.js]: We already have the functionality for free and this is a reusable component meant to be used outside of Gutenberg as well so I don't see how we gain from not exposing it.

Also, `selectedItem` and `items` very clearly convey the contract that the `selectedItem` should be an object from the `items` list.

>Also, I just noticed that the component is lacking a README.

I will work on one, but are you still unhappy with the prop names? What props would you expose?

File: storybook/test/__snapshots__/index.js.snap (hunk 9 above)
Comment (youknowriad) [reply storybook/test/__snapshots__/index.js.snap]: In my experience, each time you change the implementation, you have to change the snapshot and oftentimes you don't really check the details as it's a big blob of content.

Also, with snapshots it's not clear what are we testing exactly, we're not testing a specific behavior but just random content, it makes it hard to know what's a regression and what's not.

Another question: did you ever found aa regression with these snapshots?

File: packages/components/src/font-size-picker/index.js (hunk 10 above)
Comment (youknowriad) [reply packages/components/src/font-size-picker/index.js]: > Also, selectedItem and items very clearly convey the contract that the selectedItem should be an object from the items list.

I get that but the boat has sailed for a lot of components now, and I'd rather have users learn one way to use all of our controls rather than having each component with its own specifics.

> I will work on one, but are you still unhappy with the prop names? What props would you expose?

value, onChange and options being the important ones. 

>  I don't see how we gain from not exposing it.

we gain that if we ever change the implementation or anything, we're less likely to break backward compatibility. Adding new APIs anywhere in our codebase shouldn't be considered a "good thing", it's something to do with a lot of caution.

File: storybook/test/__snapshots__/index.js.snap (hunk 9 above)
Comment (epiqueras) [reply storybook/test/__snapshots__/index.js.snap]: I think we use them for things like highlighting missing/removed class names that plugins rely on.

File: packages/components/src/font-size-picker/index.js (hunk 10 above)
Comment (epiqueras) [reply packages/components/src/font-size-picker/index.js]: >I get that but the boat has sailed for a lot of components now, and I'd rather have users learn one way to use all of our controls rather than having each component with its own specifics.

That's fair and I can definitely see value in that if we stick with it across all inputs. I updated the prop names in this PR: #18827.

>we gain that if we ever change the implementation or anything, we're less likely to break backward compatibility. Adding new APIs anywhere in our codebase shouldn't be considered a "good thing", it's something to do with a lot of caution.

Agreed, but in this case, it's something all select implementations could easily expose, so I wouldn't worry about backwards compatibility with it. I still haven't exposed it for now to be safe, we can do it when the need for it arises.

File: storybook/test/__snapshots__/index.js.snap (hunk 9 above)
Comment (tellthemachines) [reply storybook/test/__snapshots__/index.js.snap]: My take on it is snapshots are purely for testing markup: making sure no HTML has been accidentally stripped out or made unsemantic, classname or aria attribute changes, etc. It'll only be a big blob if we make a lot of changes! In this instance it's quite big because we changed the markup for the whole component. Still, I find markup changes much more readable in this format than in JSX.

I have found regressions before with snapshots; mostly of the kind mentioned above, but sometimes I've spotted issues with semantics on components being built from scratch too, because the snapshot view is pretty close to the output HTML.

Of course snapshots are only useful when we read them, but that is just a cultural issue that we can work on improving through educating on good testing and code review practice 😀

Comment (epiqueras) [reply storybook/test/__snapshots__/index.js.snap]: Yeah, this is my take on them too.

Comment (youknowriad) [reply storybook/test/__snapshots__/index.js.snap]: It's very hard to argue against tests :) and I had this exact same conversation over and over again with different people and everytime we land the snapshots and decide to rollback later :). Nothing personal here but I think this is just one of these occasions.

Comment (tellthemachines) [reply storybook/test/__snapshots__/index.js.snap]: I'd love to know why you decided to roll back! In my previous project snapshots were pretty much universally agreed on. This PR might not be the best place for discussion 😄, but maybe we could bring it up at a Slack meeting at some point?

Comment (youknowriad) [reply storybook/test/__snapshots__/index.js.snap]: What I said above. basically after some time using them, we noticed that they are not that useful.

**Comment by afercia** [comment]
Looks like there are a few missing things in this component, will comment on the issue.

**Comment by aduth** [comment]
Noting that this Downshift dependency is fairly large at 32.4kb, contributing ~200ms additional load time over 3G.

https://bundlephobia.com/result?p=downshift@3.4.5

**Comment by silviuavram** [comment]
It is true, the bundle size is getting bigger, and I am close to release `useCombobox`. I am considering having useSelect and other hooks as part of a a11y monorepo, as this size issue popped up to us recently as well. If you'd like to help out in converging to a solution to the bundle size that would be fantastic.

It should not break the current behavior (`useSelect` being as part of `Downshift`) and also it should be part of a monorepo with a11y hooks. It would be interesting if it can be imported in Downshift from that monorepo and tree-shaken if not used. I am pretty new to these concepts so I would love some opinions. Maybe @kentcdodds can help out with an idea.

Thoughts?

**Comment by epiqueras** [comment]
>It should not break the current behavior (useSelect being as part of Downshift) and also it should be part of a monorepo with a11y hooks. It would be interesting if it can be imported in Downshift from that monorepo and tree-shaken if not used.

Sounds like a reasonable plan.

**Comment by silviuavram** [comment]
Although I am thinking ... if you are using `import` to get only `useSelect`, and I am using `rollup` to bundle, shouldn't you get only that hook, and not the whole library?

**Comment by epiqueras** [comment]
Yeah `Downshift` has an `esm` distribution. I think our tooling doesn't support it, but we can work on that right @aduth?

**Comment by aduth** [comment]
If it can be tree-shaken, our build should be able to drop the unused code. I was judging based on the [BundlePhobia page](https://bundlephobia.com/result?p=downshift@3.4.5), specifically this text under "Exports Analysis":

>Exports analysis is available only for packages that export ES Modules and are side-effect free. This package exports ES6 modules, but isn't marked side-effect free.

**Comment by epiqueras** [comment]
I guess it's not explicitly marked side-effect free, but I don't think it has any side-effects does it?

**Comment by aduth** [comment]
It still needs to be included in the published `package.json` for Webpack to consider it for tree shaking:

>The new webpack 4 release expands on this capability with a way to provide hints to the compiler via the "`sideEffects`" `package.json` property to denote which files in your project are "pure" and therefore safe to prune if unused.

https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free

**Comment by epiqueras** [comment]
It looks like it will still try to tree-shake most things, but will trip up with things like HOCs without directives:

https://webpack.js.org/guides/tree-shaking/#clarifying-tree-shaking-and-sideeffects

@silviuavram can we add the `sideEffects` hint to make this more efficient?

**Comment by silviuavram** [comment]
I am really interested to try this out, I will make changes on `Downshift` side if needed. Let me know on the progress please!

**Comment by epiqueras** [comment]
@silviuavram it's `Downshift` that is missing the `sideEffects` property, we need to add it there.

**Comment by silviuavram** [comment]
Can you create a PR and I will look over it tomorrow? Thank you! Do we need
to publish an alpha version to test if it works?

**Comment by epiqueras** [comment]
@silviuavram https://github.com/downshift-js/downshift/pull/843

>Do we need
to publish an alpha version to test if it works?

Yeah, that would be helpful. It should work, unless `Downshift` has side effects, but better to be safe.

**Comment by silviuavram** [comment]
Thanks very much! I will publish an alpha tomorrow from that branch and
test it out.

**Comment by silviuavram** [comment]
I published downshift@3.5.0-alpha.0 and could not manage to properly test it. Maybe my testing project is not configured as it should. If you have any tips that will help me looking into it tomorrow that will be super helpful.

**Comment by epiqueras** [comment]
What was the issue?

**Comment by silviuavram** [comment]
I probably did not know what to look for in the minified file. In the
main.map.js I could still find all the exports (useSelect, Downshift,
useCombobox) even though I only imported: import {useCombobox} from
‘downshift’. I probably did not configure the webpack in my test project
correctly. Or it’s something wrong in Downshift’s rollup config or
index.js. Or maybe it works and I could not figure tjat out from the
minified file. I don’t know yet.

**Comment by epiqueras** [comment]
Can you push your test somewhere?

**Comment by silviuavram** [comment]
https://github.com/silviuavram/check-downshift-shake tried it here.

**Comment by silviuavram** [comment]
I updated the repo with some things I tried on babel. Still it imports the whole things in the ES bundle.

**Comment by epiqueras** [comment]
Yeah, you're building in development mode:

https://github.com/silviuavram/check-downshift-shake/blob/27a5fcb763ce2b837017a948a6c859ccfa0a0148/package.json#L11-L12

See, https://webpack.js.org/guides/tree-shaking/.

**Comment by silviuavram** [comment]
I tried to build with —production as well. The es has the same size as the
cjs.

**Comment by epiqueras** [comment]
I just debugged this and it's probably because of how `Downshift` bundles its `esm` module.

There are a lot of static property assignments in there like `Downshift.stateChangeTypes = stateChangeTypes;` and `useCombobox.stateChangeTypes = stateChangeTypes$2;`.

Some prior discussion:

https://github.com/downshift-js/downshift/issues/380

**Comment by silviuavram** [comment]
Looks good, I was also suspicious about those state change things while going through the bundled code. Thanks so much for looking into this!

The bad part is the fact that we need to deliver a breaking change to make the bundle tree-shakeable again: exporting those `stateChangeTypes` separately and removing the static allocation. Will check if indeed, with these changes, the problem will be fixed.

**Comment by epiqueras** [comment]
There are other workarounds:

- You could break down the output into multiple files with their ESM imports intact and that would allow the `sideEffects` flag to work. Currently, Webpack evaluates the single output file, because there are used exports there, and it can't prune the unused ones because they are mutated. If they were imported and mutated from a different file, Webpack could use the `sideEffects` hint and avoid evaluating those files.

- Another, perhaps easier change, would be to use the `/*#__PURE__*/` annotations. They're like the `sideEffects` hint, but for statements instead of modules.

Related discussion: https://github.com/styled-components/babel-plugin-styled-components/issues/245.

**Comment by silviuavram** [comment]
Found a few problems with today's investigation:

The hooks were not tree-shaken because of
```js
const validatePropTypes = getPropTypesValidator(useSelect, propTypes)
```

`Downshift` component not tree-shaken because of the `static` props `stateChangeTypes` and `defaultProps!!!`. The `propTypes` static one was correctly removed by a babel transpiler (I think) at bundle time, like:
```process.env.NODE_ENV !== "production" ? Downshift.propTypes = {```

Probably I could add this same check manually to the hooks, because I don't need that prop types check function to run in production. On the `Downshift` static fields `stateChangeTypes` and `defaultProps` maybe there is a solution with one of your suggestions?

**Comment by epiqueras** [comment]
Yeah, both should work, albeit the first one is a better long-term solution that will work for future code as well.

**Comment by youknowriad** [comment]
Any plans to add a README for this component.

**Comment by epiqueras** [comment]
#19026

**Comment by adamsilverstein** [comment]
With Downshift merged, we should reconsider using it elsewhere (eg for Author selection, page parent selection, etc... eg https://github.com/WordPress/gutenberg/pull/7478). 

@afercia I missed most of the work/conversation here - does the progress on Downshift address the feedback from the original ticket? shall I work on a new PR for you to review?

**Comment by epiqueras** [comment]
Yes, that is a good next step with it.
